pkg/multicluster/watcher.go

```diff
@@ -0,0 +1,216 @@
+package multicluster
+
+import (
+	"context"
+	"sync"
+
+	"github.com/rotisserie/eris"
+	"github.com/solo-io/go-utils/contextutils"
+	"github.com/solo-io/skv2/pkg/generated/kubernetes/core/v1/controller"
+	"github.com/solo-io/skv2/pkg/multicluster/kubeconfig"
+	"github.com/solo-io/skv2/pkg/reconcile"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	// MasterCluster is the clusterName for the cluster ClusterWatcher watches.
+	MasterCluster = ""
+)
+
+// ClusterWatcher watches for KubeConfig secrets on the master cluster.
+// It is responsible for starting cluster managers and calling ClusterHandler functions.
+type ClusterWatcher interface {
+	// Run starts a watch for KubeConfig secrets on the cluster managed by the given manager.Manager.
+	// Note that Run will call Start on the given manager and run all registered ClusterHandlers.
+	Run(master manager.Manager) error
+	// RegisterClusterHandler adds a ClusterHandler to the ClusterWatcher.
+	RegisterClusterHandler(handler ClusterHandler)
+}
+
+// ManagerSet maintains a manager for every cluster in the system.
+type ManagerSet interface {
+	// Cluster returns a manager for the given cluster, or an error if one does not exist.
+	Cluster(cluster string) (manager.Manager, error)
+}
+
+type clusterWatcher struct {
+	ctx      context.Context
+	handlers *handlerList
+	managers *managerSet
+	options  manager.Options
+}
+
+var _ ClusterWatcher = &clusterWatcher{}
+var _ ManagerSet = &clusterWatcher{}
+
+// NewClusterWatcher returns a *clusterWatcher.
+// When ctx is cancelled, all cluster managers started by the clusterWatcher are stopped.
+// Provided manager.Options are applied to all managers started by the clusterWatcher.
+func NewClusterWatcher(ctx context.Context, options manager.Options) *clusterWatcher {
+	return &clusterWatcher{
+		ctx:      ctx,
+		handlers: newHandlerList(),
+		managers: newManagerSet(),
+		options:  options,
+	}
+}
+
+func (c *clusterWatcher) Run(master manager.Manager) error {
+	c.startManager(MasterCluster, master)
+	loop := controller.NewSecretReconcileLoop("cluster watcher", master)
+	return loop.RunSecretReconciler(c.ctx, c, kubeconfig.Predicate())
+}
+
+func (c *clusterWatcher) ReconcileSecret(obj *v1.Secret) (reconcile.Result, error) {
+	clusterName, cfg, err := kubeconfig.SecretToConfig(obj)
+	if err != nil {
+		return reconcile.Result{}, eris.Wrap(err, "failed to extract kubeconfig from secret")
+	}
+
+	// If the cluster already has a manager, remove the existing instance and start again.
+	if _, err := c.managers.get(clusterName); err == nil {
+		c.removeCluster(clusterName)
+	}
+
+	mgr, err := manager.New(cfg.RestConfig, c.optionsWithDefaults())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	c.startManager(clusterName, mgr)
+
+	return reconcile.Result{}, nil
+}
+
+func (c *clusterWatcher) ReconcileSecretDeletion(req reconcile.Request) {
+	// TODO update to namespace.name
```

Comment:
This vague comment alludes to updating our concept of "cluster name" from the name on the kubeconfig secret to the namespace.name on the kubeconfig secret.

While this would eliminate any risk of name duplication, it's also a bit cumbersome, and leaks the detail of how clusters are stored into all places we reference clusters. This includes code and any resource with a multicluster ref. I want to be clear on how we intend to refer to clusters in the long term before committing to this approach.

Reply:
going to remove the todo from code and file an issue about fleshing out our cluster registry story